### PR TITLE
Scope the fetch-params lock file to the user

### LINKED ISSUE #3925: Incorrectly scoped use of temporary files in fetch-params.sh
local lockfile=/tmp/fetch_params.lock

Should be done with a secure temporary name.

### PR DESCRIPTION
**Pull request by sellout**
Also takes the opportunity to use more standard location for the lockfile and
improves the lock-conflict error to help a user if the fetcher crashed without
removing the lock or if they are running it from a distribution in which it’s
called `zcash-fetch-params` instead of `fetch-params.sh`.

Fixes #3925

### CODE REVIEW

**Review by nuttycom — DISMISSED** [review]

File: zcutil/fetch-params.sh [hunk 1]
```diff
@@ -1,8 +1,15 @@
-#!/bin/sh
+#!/usr/bin/env bash
 
 export LC_ALL=C
 set -eu
 
+SCRIPT_NAME=$(basename $0)
+
+[[ -v XDG_CACHE_DIR ]] || XDG_CACHE_DIR="${HOME}/.cache"
```
[the remark this answers is not in the document]
Comment (sellout) [reply zcutil/fetch-params.sh]: 🤦

I ended up testing it this time by making the fallback non-standard, and then running `unset XDG_RUNTIME_DIR; ./zcutil/fetch-params.sh` and making sure that the lock file ended up in the expected (standard) cache location. Annoyingly, `unset` doesn’t care what shell you’re in, and now my whole Emacs session seems b0rked 😅

File: zcutil/fetch-params.sh [hunk 2]
```diff
@@ -199,10 +210,10 @@ EOF
         # This may be the first time the user's run this script, so give
         # them some info, especially about bandwidth usage:
         cat <<EOF
-The complete parameters are currently just under 1.7GB in size, so plan 
+The complete parameters are currently just under 1.7GB in size, so plan
 accordingly for your bandwidth constraints. If the Sprout parameters are
```
Comment (daira) [inline zcutil/fetch-params.sh:214]:
```suggestion
The complete parameters are currently just under 800MB in size, so plan
accordingly for your bandwidth constraints.
```

File: zcutil/fetch-params.sh [hunk 3]
```diff
@@ -199,10 +210,10 @@ EOF
...
-already present the additional Sapling parameters required are just under 
-800MB in size. If the files are already present and have the correct 
+already present the additional Sapling parameters required are just under
+800MB in size. If the files are already present and have the correct
 sha256sum, no networking is used.
```
Comment (daira) [inline zcutil/fetch-params.sh:217]:
```suggestion
present and have the correct sha256sum, no networking is used.
```

File: zcutil/fetch-params.sh [hunk 4]
```diff
@@ -225,7 +236,7 @@ EOF
     fetch_params "$SAPLING_SPROUT_GROTH16_NAME" "$PARAMS_DIR/$SAPLING_SPROUT_GROTH16_NAME" "b685d700c60328498fbde589c8c7c484c722b788b265b72af448a5bf0ee55b50"
```
Comment (daira) [inline zcutil/fetch-params.sh:236]: Consider changing the Sprout part to
```shell
# Delete Sprout parameters:
delete_params "$PARAMS_DIR/$SPROUT_PKEY_NAME" "8bc20a7f013b2b58970cddd2e7ea028975c88ae7ceb9259a5344a16bc2c0eef7"
delete_params "$PARAMS_DIR/$SPROUT_VKEY_NAME" "4bd498dae0aacfd8e98dc306338d017d9c08dd0918ead18172bd0aec2fc5df82"
```
where
```shell
delete_params() {
    # We only set these variables inside this function,
    # and unset them at the end of the function.
    output="$1"
    expectedhash="$2"
    if [ -f "$output" ]
    then
        "$SHA256CMD" $SHA256ARGS -c <<EOF
$expectedhash  $output
EOF
        # Check the exit code of the shasum command:
        CHECKSUM_RESULT=$?
        if [ $CHECKSUM_RESULT -eq 0 ]; then
            rm -f "$output"
        else
            echo "Checksum of old Sprout parameter file ${output} is not as expected -- not deleting!" >&2
            exit 1
        fi
    fi
    unset -v output
    unset -v expectedhash
}
```

**Review by daira — DISMISSED** [review]
utACK with suggestions.

File: zcutil/fetch-params.sh [hunk 5]
```diff
@@ -225,7 +234,7 @@ EOF
...
 }
 
-if [ "x${1:-}" = 'x--testnet' ]
+if [ "${1:-}" = '--testnet' ]
```
Comment (daira) [inline zcutil/fetch-params.sh:237]: Yes the rationale for the "x" thing is wrong; there's never any ambiguity for `test "$a" = 'b'` or `[ "$a" = 'b' ]`: https://stackoverflow.com/questions/174119/why-do-shell-script-comparisons-often-use-xvar-xyes/174288#comment54562572_180630

**Review by daira — APPROVED** [review]
utACK

File: zcutil/fetch-params.sh (hunk 4 above)
Comment (daira) [reply zcutil/fetch-params.sh:234]: I'll file a separate issue for this.

**Comment by daira** [comment]
The mode change 100755 → 100644 on `fetch-params.sh` is incorrect I think.

**Comment by sellout** [comment]
> The mode change 100755 → 100644 on `fetch-params.sh` is incorrect I think. It should be executable.

That’s odd – and so subtle in the diff! Thanks for catching it!

**Comment by sellout** [comment]
Ah, seems like a consequence of resolving the merge conflict with master via the Web UI.

File: zcutil/fetch-params.sh (hunk 4 above)
Comment (sellout) [reply zcutil/fetch-params.sh]: #6295

**Review by daira — APPROVED** [review]
utACK

**Review by str4d — APPROVED** [review]
utACK 4cd627a9857aa945974ea48ba30fabad2fc4ea33

**Pull request merged**